### PR TITLE
Feature Request: Enable Local Model Caching when pulling from remote source like s3

### PR DESCRIPTION
**Pull request by wwvela**
### Feature Description

Adds write-through local caching for model files streamed from object storage. When enabled, tensor data is written to a local directory during streaming (zero-copy via buffer protocol) and served from local filesystem on subsequent loads. According Issue for this feature discussion: https://github.com/run-ai/runai-model-streamer/issues/152

### Changes

**New files:**
- `py/runai_model_streamer/runai_model_streamer/cache/__init__.py`
- `py/runai_model_streamer/runai_model_streamer/cache/cache.py` — `StreamCache` class with TP-aware cache keys, atomic writes, race condition handling
- `docs/src/local-cache.md` — feature documentation

**Modified files:**
- `py/runai_model_streamer/runai_model_streamer/__init__.py` — logging handler (opt-in via `RUNAI_STREAMER_LOG_LEVEL`)
- `py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py` — cache integration (hit/miss logic, write-through, `enable_cache` parameter)
- `py/runai_model_streamer/runai_model_streamer/distributed_streamer/distributed_streamer.py` — pass `enable_cache`, call `set_distributed(rank, world_size)` for TP-aware keys
- `py/runai_model_streamer/runai_model_streamer/safetensors_streamer/safetensors_streamer.py` — pass `enable_cache=True` for tensor data stream (not header reads)
- `docs/src/env-vars.md` — document `RUNAI_STREAMER_CACHE_DIR`

### How It Works

**First load (cache miss):**
```
S3 → C++ streamer → DRAM buffer → GPU
                          │
                          └→ os.write() to /cache/<file>.partial.<pid>
                             (inline, buffer protocol, no extra copy)
                             When file complete: rename → final, write .done sentinel
```

**Second load (cache hit):**
```
/cache/<file> → C++ streamer (local filesystem) → GPU
(NVMe speed, no network)
```

### Design Decisions

1. **Write-through, not write-back**: Data is written to cache inline during streaming. No background thread, no second S3 download. The only overhead is `os.write()` to page cache which is typically non-blocking.

2. **TP-aware cache keys**: With distributed streaming, each worker caches its own tensor partition. Cache key includes rank and world_size: `model-00001.safetensors.tp4_rank0.<hash>`. Same TP config → cache hit; different TP → cache miss (no corruption).

3. **Atomic writes**: Data goes to `.partial.<pid>` temp file, then `os.rename()` to final path. The `.done` sentinel is only written after rename succeeds. Prevents serving incomplete files if process dies mid-write.

4. **Race condition handling**: Multiple TP workers (separate processes) may cache the same file. Each writes to its own `.partial.<pid>`. First to finish renames to final; others detect existing file and clean up. No locks needed.

5. **All-or-nothing cache hit**: The C++ layer cannot mix local and remote paths in one request. Cache is used only when ALL files for a worker hit. Partial hits fall back to full S3 streaming (while scheduling background cache writes for misses).

6. **Only tensor data is cached**: The `enable_cache=True` flag is only passed for the tensor data stream call, not the header reads (8 bytes + JSON). This prevents tiny header reads from creating invalid cache entries.

7. **Short write handling**: `os.write()` on Linux can return less than requested for buffers >2GB (`ssize_t` limit). The write loop retries until all bytes are written.

### Configuration

```yaml
env:
  - name: RUNAI_STREAMER_CACHE_DIR
    value: "/models"
volumeMounts:
  - name: model-cache
    mountPath: /models
volumes:
  - name: model-cache
    hostPath:
      path: /mnt/k8s-disks/0/models
      type: DirectoryOrCreate
```

### Cache Directory Structure

```
/models/
├── model-00001.safetensors.tp4_rank0.61cc262a76817ddb
├── model-00001.safetensors.tp4_rank0.61cc262a76817ddb.done
├── model-00001.safetensors.tp4_rank1.61cc262a76817ddb
├── model-00001.safetensors.tp4_rank1.61cc262a76817ddb.done
└── ...
```

Without distributed streaming (TP=1):
```
/models/
├── model-00001.safetensors.61cc262a76817ddb
├── model-00001.safetensors.61cc262a76817ddb.done
└── ...
```

### Sentinel File (.done)

```json
{
  "remote_path": "s3://bucket/model/model-00001.safetensors",
  "file_offset": 8392,
  "size": 1073741824,
  "rank": 0,
  "world_size": 4
}
```

### Testing

- Unit tests: `test_cache.py` covers cache enable/disable, hit/miss, write-through, race conditions
- Integration tested with vLLM on:
  - Ministral-3-8B (22GB) on g6e.12xlarge (TP=2, TP=4)
  - Qwen3.6-35B-A3B (64GB) on g6e.12xlarge (TP=4) and p5.48xlarge (TP=4, TP=8)
  - Llama-4-Scout-17B-16E (220GB) on p5.48xlarge (TP=4, TP=8)

**g6e.12xlarge (4x L40S 48GB, 40 Gbps NIC, single NVMe ~3.5 GB/s, 96GB RAM)**

| Model | TP | S3 (no cache) | NVMe preloaded (s5cmd) | Cache 1st pod (cache miss) | Cache 2nd pod (NVMe cache hit) |
|-------|-----|--------------|----------------------|---------------|--------------------------|
| Ministral-3-8B (22GB) | 2 | 8.5s | 6.3s | 8.3s (~same as S3) | 6.9s |
| Qwen3.6-35B-A3B (64GB) | 4 | 29.6s | 20.0s | 31.7s* (+2s) | 21.1s |

*With `dirty_ratio=80` node tuning for Qwen3.6-35B-A3B case. Without tuning: 43.7s (+14s overhead). See [Kernel Dirty Page Pressure](#kernel-dirty-page-pressure).

**p5.48xlarge (8x H100 80GB, 200 Gbps NIC, 4x NVMe striped ~14 GB/s, 2TB RAM)**

| Model | TP | S3 (no cache) | NVMe preloaded (s5cmd) | Cache 1st pod (cache miss) | Cache 2nd pod (NVMe cache hit) |
|-------|-----|--------------|----------------------|---------------|--------------------------|
| Qwen3.6-35B-A3B (64GB) | 4 | 15.7s | 7.0s | 18.7s (+3s) | 7.6s |
| Llama-4-Scout (220GB) | 4 | 33.6s | 9.9s | 29.4s (~same) | 9.6s |

**Key observations:**
- Cache 2nd pod performance matches NVMe preloaded (s5cmd daemonset) — same local NVMe read speed
- First-pod cache write overhead is negligible on high-RAM instances (p5) and for small models
- Cache hit consistently provides 2-7x faster weight loading vs S3
- Larger models benefit more from caching (220GB: 71% improvement on scale-up)
- The cache approach eliminates the need for a preload daemonset while achieving equivalent NVMe read performance

### Compatibility

- Backward compatible: no behavior change when `RUNAI_STREAMER_CACHE_DIR` is not set
- Works with and without distributed streaming
- Expected to work with all object storage backends (S3, GCS, Azure), but only test with s3 above
- Cache entries are isolated by TP config — changing TP causes cache miss, not corruption
- No C++ changes required

### Known Limitations

- Cache hit requires ALL files to be cached (partial hits fall back to S3)
- Cache is tied to TP configuration (TP=4 cache cannot serve TP=8)
- On instances with limited RAM, large model cache writes may hit kernel dirty page limits (see below #Kernel Dirty Page Pressure)  — recommend setting `vm.dirty_ratio=80` on such nodes

### Kernel Dirty Page Pressure

Linux's `dirty_ratio` (default 20% of RAM) controls when `os.write()` blocks. When dirty pages exceed this threshold, the kernel forces synchronous writeback — stalling the streaming pipeline until pages flush to disk.

**How the limit is hit:**

With distributed streaming (TP=N), each worker writes its partition to cache simultaneously:
```
Total dirty pages generated = (model_size / TP) × TP = model_size
All workers write concurrently → all dirty pages accumulate at once
```

The write blocks when: `model_size > dirty_ratio × instance_RAM`

**Examples:**

| Instance | RAM | Model | Dirty pages generated | 20% limit | Exceeds? | Impact |
|----------|-----|-------|----------------------|-----------|----------|--------|
| g6e.12xlarge | 96GB | 64GB (TP=4) | 16GB × 4 = 64GB | 19GB | **3.4x over** | +16.6s overhead |
| g6e.12xlarge | 96GB | 22GB (TP=4) | 5.5GB × 4 = 22GB | 19GB | 1.2x over | ~0s (writes interleave with downloads) |
| g6e.48xlarge | 384GB | 64GB (TP=4) | 16GB × 4 = 64GB | 77GB | No | No impact |
| p5.48xlarge | 2TB | 64GB (TP=4) | 16GB × 4 = 64GB | 400GB | No | No impact |
| p5.48xlarge | 2TB | 220GB (TP=8) | 27.5GB × 8 = 220GB | 400GB | No | No impact |

**Rule of thumb:** Cache write overhead is significant when `model_size > 20% of instance RAM`.

**Fix:** Set `vm.dirty_ratio=80` on affected nodes:
```bash
echo 80 > /proc/sys/vm/dirty_ratio
echo 50 > /proc/sys/vm/dirty_background_ratio
```

**Results (g6e.12xlarge, 64GB model):**

| | dirty_ratio=20% (default) | dirty_ratio=80% |
|---|---|---|
| First load with cache | 43.7s | 31.7s |
| Cache write overhead | +16.6s | +4.6s |
| Remaining overhead source | Physical NVMe write speed (~3.5 GB/s) | Same (unavoidable) |

This tuning is only needed for instances where `model_size > 20% × RAM`. For most production deployments on p4d/p5 class instances (1-2TB RAM), the default is sufficient. But the model cache actually shouldn't be used when using a instance type with relative small RAM size compared with model size, so this kernel Dirty Page Pressure might not be a real issue here.

## Summary

* **New Features**
  * Added opt-in local filesystem caching for streamed files, reusing fully cached artifacts on subsequent requests.
  * Introduced an `enable_cache` option across distributed streaming and safetensors streaming for rank-aware cache usage.

* **Documentation**
  * Added a guide covering cache directory/layout, cache miss vs hit behavior, and tuning/logging for concurrency and performance.

* **Bug Fixes**
  * Improved cache correctness by verifying cache completion metadata before reuse and cleaning up incomplete cache artifacts on errors/aborts.

### CODE REVIEW

File: py/runai_model_streamer/runai_model_streamer/cache/cache.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            self._fd = -1
+            # Another worker may have already written the final file
+            if os.path.exists(self._final_path):
+                logger.info(f"[RunAI Streamer][Cache] Already cached by another worker: {self._remote_path}")
+                self._cleanup()
+                return
+            os.rename(self._tmp_path, self._final_path)
+            meta = {"remote_path": self._remote_path, "file_offset": self._file_offset, "size": self._written, "rank": self._rank, "world_size": self._world_size}
+            with open(self._sentinel, "w") as f:
+                json.dump(meta, f)
```
[the remark this answers is not in the document]
Comment (wwvela) [reply py/runai_model_streamer/runai_model_streamer/cache/cache.py:73]: Intentionally omitted. We tested with os.fsync(self._fd) before close and it caused a 10x performance regression (80s vs 8s for a 22GB model) because fsync serializes the streaming pipeline — each file's fsync blocks until multi-GB of data physically flushes to NVMe.

The tradeoff we chose: this cache is a performance optimization, not a durability store. The source of truth is always S3. If a power loss corrupts a cached file, the worst case is a cache miss on next load (re-download from S3). The .done sentinel includes the expected size field, so a future enhancement could add a file-size validation on cache hit to detect truncation without paying the fsync cost on every write.

In practice, the kernel flushes dirty pages within 30s (default dirty_expire_centisecs), and model loading pods typically stay alive long after loading completes — so the window for data loss is very small.

File: py/runai_model_streamer/runai_model_streamer/cache/cache.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    return is_s3_path(path) or is_gs_path(path) or is_azure_path(path)
+
+
+def _cache_key(remote_path: str, rank: int = 0, world_size: int = 1) -> str:
+    """Deterministic cache filename from a remote URI and TP config."""
+    h = hashlib.sha256(remote_path.encode()).hexdigest()[:16]
+    basename = os.path.basename(remote_path.rstrip("/"))
+    if world_size > 1:
+        return f"{basename}.tp{world_size}_rank{rank}.{h}"
+    return f"{basename}.{h}"
```
[the remark this answers is not in the document]
Comment (wwvela) [reply py/runai_model_streamer/runai_model_streamer/cache/cache.py:30]: The cache key already encodes the uniqueness: {basename}.tp{world_size}_rank{rank}.{sha256_of_full_uri}. For a given (remote_path, rank, world_size), the byte range is always deterministic — the distributed partition algorithm produces the same tensor slices for the same model + TP config. If TP changes, the key changes (different tp/rank prefix), causing a clean cache miss.

The only scenario where offset/size could differ for the same key is if the remote object content changed while keeping the same URI. However, model shard files in object storage are effectively immutable — they are not overwritten in place. If a user does replace a model at the same S3 path (which is unusual), they should clear the cache directory (rm -rf /models/*) to invalidate stale entries. Adding per-hit sentinel parsing for offset/size validation would cost 200+ file reads at startup for a scenario that doesn't occur in normal model serving workflows.

File: py/runai_model_streamer/runai_model_streamer/cache/cache.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wwvela) [reply py/runai_model_streamer/runai_model_streamer/cache/cache.py:73]: fix it by treat read failure as a cache miss.

File: py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py [hunk 3]
```diff
@@ -89,21 +95,64 @@ def stream_files(
...
+                self._cache_written_bytes[path] = 0
+
+        # Check cache: only use cached paths if ALL unique files hit cache (the C++ layer
+        # does not support mixed local/remote paths in a single request).
+        use_cache = enable_cache and self._cache.enabled
+        unique_paths = list(dict.fromkeys(self._cache_original_paths))
+        all_cached = use_cache and all(
+            self._cache.cached_path_and_offset(p) is not None
+            for p in unique_paths
+        )
```
[the remark this answers is not in the document]
Comment (wwvela) [reply py/runai_model_streamer/runai_model_streamer/file_streamer/file_streamer.py]: Same reasoning as the byte-range concern above — for a given (remote_path, rank, world_size), the byte layout is deterministic. The distributed partition algorithm always produces the same tensor slices in the same order for the same model + TP config. The cache key encodes all three variables, so a layout mismatch requires either a model content change at the same URI (not a normal workflow) or a TP config change (which changes the key → cache miss).

The cumulative offset tracker in the read path reconstructs the same layout that the write path used, since both derive from the same file_stream_requests order which is deterministic for a given partition.

**Comment by wwvela** [comment]
@coderabbitai resume review

File: py/runai_model_streamer/runai_model_streamer/cache/cache.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    def __init__(self, cache_dir: str, remote_path: str, file_offset: int, rank: int = 0, world_size: int = 1) -> None:
+        self._remote_path = remote_path
+        self._file_offset = file_offset
+        self._rank = rank
+        self._world_size = world_size
+        key = _cache_key(remote_path, rank, world_size)
+        self._final_path = os.path.join(cache_dir, key)
+        self._sentinel = self._final_path + ".done"
+        self._tmp_path = self._final_path + f".partial.{os.getpid()}"
+        self._fd = os.open(self._tmp_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
```
[the remark this answers is not in the document]
Comment (wwvela) [reply py/runai_model_streamer/runai_model_streamer/cache/cache.py:45]: Fix it by adding wraps the _CacheWriter(...) constructor call in open_writer with a try/except, so the os.open() failure inside __init__ is caught at the caller level and logged as a warning instead of crashing.

**Comment by wwvela** [comment]
@noa-neria Hi Noa. Addressed all ai bot comments. Would you be able to help review this PR and see if it looks good to you? Thank you!

**Comment by wwvela** [comment]
Hi @noa-neria @omer-dayan. Kindly asking for repo maintainers' ideas about this PR. Thank you

**Comment by noa-neria** [comment]
Hi @wwvela
Sorry for the delayed response! First off, thank you for putting this together—the benchmarks look great and clearly demonstrate the value of this caching feature.

I’m going to need a bit more time to give the code a thorough review, but I wanted to share a few initial points and architectural considerations to help guide the next steps:

**Technical Considerations**

- **Replica Concurrency (Mandatory):** We must handle the scenario where multiple replicas attempt to load the same files simultaneously. The current implementation doesn't protect this critical section. For local disks, `flock` is a good solution, though it won't suffice for cluster-wide PVCs.
- **Cache Management:** Because model files can scale into hundreds of gigabytes, cache size will grow rapidly. User should be aware of that and maintain cleanup strategy. For a start, a limit can be configured as `RUNAI_STREAMER_CACHE_MAX_GB`  
- **Kernel Dirty Page Limits:** To avoid performance bottlenecks, we should handle cache writes entirely in the background rather than in the yield path. However, note that this only mitigates the issue if `RUNAI_STREAMER_MEMORY_LIMIT=-1 `(unlimited). If the streamer memory buffer is limited and reused, the kernel will still block once the dirty page limit is hit.
- **Eliminating TP Dependency:** We can remove the dependency on Tensor Parallelism (TP) by setting `RUNAI_STREAMER_PARTITION_POLICY="files"` whenever the cache is enabled. This policy cleanly divides the workload at the granularity of full files.

**Architecture & Next Steps**

- **Refactoring:** To keep the codebase clean, it would be better to implement this logic in a new `FileStreamerWithCache` class extending or wrapping the existing `FileStreamer`, rather than baking it directly into the base class.
- **Testing:** Please ensure we add comprehensive test coverage for the new caching behavior and edge cases (like the concurrent loading mentioned above). Cache correctness should be verfied against the source.

Thanks again for driving this effort! Let me know your thoughts on these points.

**Comment by wwvela** [comment]
Hi Noa. Sorry for the late reply, was distracted by other works recently. And thank you so much for the comments! That's really helpful. I'm working on making changes based on them and will post new changes soon
